command/sync.go

```diff
@@ -0,0 +1,495 @@
+			dstName = filepath.ToSlash(dstObject.URL.Relative())
+		}
+
+		if srcObject == nil && dstObject == nil {
+			break
+		}
+
+		if srcObject != nil && dstObject != nil {
+			if srcName > dstName {
+				srcObject = nil
+				iSrc--
+			} else if srcName == dstName { // if there is a match.
+				commonObj = append(commonObj, &CommonObject{src: srcObject, dst: dstObject})
+			} else {
+				dstObject = nil
+				iDst--
+			}
+		}
+
+		switch {
+		case srcObject == nil && dstObject == nil:
+			// do nothing
+		case srcObject == nil:
+			dstOnly = append(dstOnly, dstObject.URL)
+		case dstObject == nil:
+			srcOnly = append(srcOnly, srcObject.URL)
+		}
+	}
+	return
+}
+
+// GetSourceAndDestinationObjects returns source and destination objects from given urls.
+func (s Sync) GetSourceAndDestinationObjects(ctx context.Context, srcurl, dsturl *url.URL) ([]*storage.Object, []*storage.Object, error) {
+	sourceClient, err := storage.NewClient(ctx, srcurl, s.storageOpts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	destClient, err := storage.NewClient(ctx, dsturl, s.storageOpts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var sourceObjects []*storage.Object
+	var destObjects []*storage.Object
+	var wg sync.WaitGroup
+
+	// get source objects.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		srcObjectChannel := sourceClient.List(ctx, srcurl, false)
+		for srcObject := range srcObjectChannel {
+			if s.shouldSkipObject(srcObject, true) {
+				continue
+			}
+			sourceObjects = append(sourceObjects, srcObject)
+		}
+	}()
+
+	// add * to end of destination string, to get all objects recursively.
+	var destinationURLPath string
+	if strings.HasSuffix(s.dst, "/") {
+		destinationURLPath = s.dst + "*"
+	} else {
+		destinationURLPath = s.dst + "/*"
+	}
+
+	destObjectsURL, err := url.New(destinationURLPath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// get destination objects.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		destObjectsChannel := destClient.List(ctx, destObjectsURL, false)
+		for destObject := range destObjectsChannel {
+			if s.shouldSkipObject(destObject, false) {
+				continue
+			}
+			destObjects = append(destObjects, destObject)
+		}
+	}()
+
+	// wait until source and destination objects are fetched.
+	wg.Wait()
+	return sourceObjects, destObjects, nil
+}
+
+// PlanRun prepares the commands and passes it to reader for run command.
+func (s Sync) PlanRun(ctx context.Context, onlySource, onlyDest []*url.URL, common []*CommonObject,
+	dsturl *url.URL, transferManager transfer.Manager, strategy strategy.Strategy, w *io.PipeWriter) {
+	// only source objects.
+	for _, srcurl := range onlySource {
+		var dsturl_ *url.URL
+		switch {
+		case !srcurl.IsRemote() && dsturl.IsRemote(): // local->remote
+			dsturl_ = transferManager.PrepareRemoteDestination(srcurl, dsturl)
+		case srcurl.IsRemote() && !dsturl.IsRemote(): // remote->local
+			dsturl_, _ = transferManager.PrepareLocalDestination(ctx, srcurl, dsturl)
```

Comment:
how about we delete  `prepareRemoteDestination` or `prepareLocalDestination` and just pass the `dsturl` which comes from the parameter. The reason for that we pass generated url to `cp` command but `cp` also calculates the destination url and there is an error handling.

Reply:
I tried that, but I guess we need to handle the remote destination. For the error handling part, I think we can skip url's with errors, what do you think?

Reply:
I think `transfer` package is not needed. As you pointed above `cp` command should create destination files/directories if destination url passed correctly. It should be enough to calculate destination urls in `sync` command then delegate it to `cp`.

Reply:
Your suggestion about the transfer package sounds good to me. But I could not understand your second comment. Why not we cannot handle errors for remote URLs? Could you refactor this part as you suggested?

Reply:
When we did not prepare the destination, `cp` commands generates destination in different manner. It is best explained with an example
Supoose that we sync files from s3 to local  and 
We run the command 
```
s5cmd sync s3://bucket/* folder/
```
one of the files in the s3 is, `s3://bucket/a/file.txt`.
1. if we handle the destination url prepation in sync, then the command becomes
```
cp s3://bucket/a/file.txt folder/a/file.txt
```
2. if we do not handle destination url prepation in sync, (cp handles it), then the command becomes
```
cp s3://bucket/a/file.txt folder/file.txt
```
it does not create the a folder, since specific url path is passed. 

Therefore, some of the tests fail right now, I can fix them, but is this behaviour wanted? What should be the correct implementation? After discussing that, we can handle the destination prepation in `sync` or `cp`. 

For the error handling part, 
`prepareLocalDestination` returns error which indicates the local prepation is successful or not, my suggestion is that, (it only applies we use `transfer` package and handle the destination prepation in `sync` ), if there is an error, we can skip or return that error in the end. What is your suggestion about that?

Reply:
👍 Thanks for the clarification.  It makes sense now. I think the first behavior is the expected one. 

```
s5cmd sync s3://bucket/* folder/
```
Should copy the file into ` folder/a/file.txt`
```
cp s3://bucket/a/file.txt folder/a/file.txt
```
Since we expand wildcard(`*`) strings in sync and pass raw URLs to run command, URL preparation should be handled in sync too.  Otherwise, destination/source URLs will not be correct as you pointed out. 


>  For the error handling part,
prepareLocalDestination returns error which indicates the local prepation is successful or not, my suggestion is that, (it only applies we use transfer package and handle the destination prepation in sync ), if there is an error, we can skip or return that error in the end. What is your suggestion about that?

Can we use multierror pattern which is used in other commands? We can log errors and append them to multierror. At the end, we can return error.

Reply:
Now it appends error at the end.